### PR TITLE
[Reviewer: Chris] Fix python packaging

### PR DESCRIPTION
**Pull request by richardwhiuk**
- Use wheels instead of eggs
- Use requirements.txt to track dependencies instead of setup.py

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -58,19 +58,23 @@ $(ENV_DIR)/bin/python:
 $(ENV_DIR)/bin/coverage: $(ENV_DIR)/bin/python
 	$(ENV_DIR)/bin/pip install coverage
 
-.PHONY: build_common_egg
-build_common_egg: $(ENV_DIR)/bin/python setup.py libclearwaterutils.a
-	$(COMPILER_FLAGS) ${ENV_DIR}/bin/python setup.py bdist_egg -d $(EGG_DIR)
+.PHONY: build_common_wheel
```
Comment (chris-elford-metaswitch) [inline Makefile]: Is the naming right? It looks like this doesn't just build the wheel for the current dir, it also builds wheels for all of the dependencies.

File: Makefile [hunk 2]
```diff
@@ -58,19 +58,23 @@ $(ENV_DIR)/bin/python:
...
-${ENV_DIR}/.eggs_installed : $(ENV_DIR)/bin/python setup.py $(shell find metaswitch -type f -not -name "*.pyc") libclearwaterutils.a
-	# Generate .egg files for python-common
-	$(COMPILER_FLAGS) ${ENV_DIR}/bin/python setup.py bdist_egg -d .eggs
+${ENV_DIR}/.wheels_installed : $(ENV_DIR)/bin/python setup.py requirements.txt $(shell find metaswitch -type f -not -name "*.pyc") libclearwaterutils.a
+	rm -rf .wheelhouse
 
-	# Download the egg files they depend upon
-	${ENV_DIR}/bin/easy_install -zmaxd .eggs/ .eggs/*.egg
+	# Generate .whl files for python-common and dependencies
+	$(COMPILER_FLAGS) ${ENV_DIR}/bin/pip wheel -w .wheelhouse -r requirements.txt .
```
Comment (chris-elford-metaswitch) [inline Makefile]: This line is almost the same as the `build_common_wheel` target. Why are they so similar but different? Shouldn't we use the wheels in the `WHEELHOUSE` for install and test?

File: Makefile [hunk 3]
```diff
@@ -87,7 +91,7 @@ pyclean:
 
 .PHONY: envclean
 envclean:
-	rm -rf bin .eggs .develop-eggs parts .installed.cfg bootstrap.py .downloads .buildout_downloads
+	rm -rf bin .wheelhouse .wheels_installed .develop-eggs parts .installed.cfg bootstrap.py .downloads .buildout_downloads
```
Comment (chris-elford-metaswitch) [inline Makefile]: What command cleans up `${WHEELHOUSE}`?

File: Makefile (hunk 2 above)
Comment (chris-elford-metaswitch) [reply Makefile]: Ah, I see. It's not clear to me from this Makefile how we expect `build_common_wheel` to be used. I think it should be possible to commonize some code here?

File: Makefile (hunk 3 above)
Comment (chris-elford-metaswitch) [reply Makefile]: Ah, it doesn't exist. Cool

Comment (richardwhiuk) [reply Makefile]: To be clear, `${WHEELHOUSE}` belongs to the calling package, so it's the calling packages responsibility for cleaning it up.

File: Makefile (hunk 1 above)
Comment (richardwhiuk) [reply Makefile]: Correct. On reflection, I think it should just build the common wheel, and then we end up with a more straight forward approach.

File: Makefile (hunk 2 above)
Comment (richardwhiuk) [reply Makefile]: I've tweaked the above to just build itself, whereas this needs to build the wheel and be ready to install all of it's dependencies.

**Pull request merged**